contracts/swappers/BalancerSwapper.sol

```diff
@@ -0,0 +1,177 @@
+import "../interfaces/IBVault.sol";
+import "../interfaces/IBWeightedPoolMinimal.sol";
+import "../proxy/ControllableV3.sol";
+import "../openzeppelin/Math.sol";
+import "../lib/WeightedMath.sol";
+
+/// @title Swap tokens via Balancer vault.
+/// @notice !!! Weighted pools only supported yet !!!
+/// @author bogdoslav
+contract BalancerSwapper is ControllableV3, ISwapper {
+  using SafeERC20 for IERC20;
+  address public balancerVault;
+
+  // *************************************************************
+  //                        CONSTANTS
+  // *************************************************************
+
+  /// @dev Version of this contract. Adjust manually on each code modification.
+  string public constant BALANCER_SWAPPER_VERSION = "1.0.0";
+  uint public constant PRICE_IMPACT_DENOMINATOR = 100_000;
+
+  uint private constant _ASSET_IN_INDEX = 0;
+  uint private constant _ASSET_OUT_INDEX = 1;
+  uint private constant _LIMIT = 1;
+
+  // *************************************************************
+  //                        VARIABLES
+  //                Keep names and ordering!
+  //                 Add only in the bottom.
+  // *************************************************************
+
+
+  // *************************************************************
+  //                        EVENTS
+  // *************************************************************
+
+  event Swap(
+    address pool,
+    address tokenIn,
+    address tokenOut,
+    address recipient,
+    uint priceImpactTolerance,
+    uint amountIn,
+    uint amountOut
+  );
+  // *************************************************************
+  //                        INIT
+  // *************************************************************
+
+  /// @dev Proxy initialization. Call it after contract deploy.
+  function init(address controller_, address balancerVault_) external initializer {
+    __Controllable_init(controller_);
+    require(balancerVault_ != address(0), 'Zero balancerVault');
+    balancerVault = balancerVault_;
+  }
+
+  // *************************************************************
+  //                     GOV ACTIONS
+  // *************************************************************
+
+  // *************************************************************
+  //                        PRICE
+  // *************************************************************
+
+  function getPrice(
+    address pool,
+    address tokenIn,
+    address tokenOut,
+    uint amount
+  ) public view override returns (uint) {
+    { // take pool commission
+      uint swapFeePercentage = IBWeightedPoolMinimal(pool).getSwapFeePercentage();
+      amount -= amount * swapFeePercentage / 10**18;
+    }
+    bytes32 poolId = IBWeightedPoolMinimal(pool).getPoolId();
+    (IERC20[] memory tokens,
+    uint256[] memory balances,) = IBVault(balancerVault).getPoolTokens(poolId);
+
+    require(tokens.length == 2, 'Wrong pool tokens length');
```

Comment:
why we can not use it for pools with multiple tokens?

Reply:
It is compatible. Unnecessary check removed. Code improved.